### ISSUE TITLE
Parameter erroneously marked as unused

### ISSUE THREAD

**Issue by InSyncWithFoo** [microsoft/pyright#7772]
Minimal reproducible example ([playground](https://pyright-play.net/?pythonVersion=3.13&strict=true&code=GYJw9gtgBAxmA28CmMAuBLMA7AzgOgEMAjGKdCABzBFSgGEDFjkBYAKHZngJx3oC52UYVAo8c7dqgCeFJFABiAbQAqARgA0UFQCYAulAC89RtyLIlqtXq26DAHxNNzSS%2BtvXb%2ByWyL8nZhZKdDb0BsbcEEQAJgRQBFp%2B8VAA1FBEIplZ2TkAxFBIIOAg-gBEABQE-nSJ-gCqWADWWGAA7lgAlFAAtAB8UA3NbVilZHzoWHCUBBguUK3oqAAWUBV0XX30pexVAcyuwVohR%2BFQkTFxCen%2BcWkZOQ-Z%2BYXFZSmjLbQ4AK4UVDRIaJQYDUVZ0UYELBA0rgnzANT%2BZQ1MJGM4EKKxeK1ZJ3R54-EifItApFahaAAGBHJ8ShUHJRGpEAIIEagPifG%2BWG%2BOEB7GAOmqpn2lmOKMcDGcQWRopCqPOmKuSVu6WERLAJOKFKpNKB9MZzNZQJ4UE53N5bGAAGZEUoeaglBNUKE7Q6sE7TvLLtjlfcCVk1RqyXTtZDdQyoEyWWzjaaedF2EA)):

(Mostly copied from [this Stack Overflow question](https://stackoverflow.com/q/78387312).)

```py
from collections.abc import Callable

class C:
    pass

type F[T1, T2] = Callable[[T1], T2] | Callable[[T1, T1], T2]

b: Callable[[C], C] = lambda a, b: a + b                         # error: "(a: C, b: Unknown) -> Unknown" is incompatible with "(C) -> C"
a: Callable[[C, C], C] = lambda a, b: a + b                      # error: "+" not supported for "C" and "C"

f1: F[C, C] = lambda a, b: a + b                                 # no error, `a` and `b` marked as unused
f2: Callable[[C], C] | Callable[[C, C], C] = lambda a, b: a + b  # no error, `a` and `b` marked as unused
f3: F[set[int], set[int]] = lambda a, b: a + b                   # no error, `a` and `b` marked as unused
```

I expected all assignments to be marked as errors, since `C` doesn't support addition, and parameters not marked as unused.

**Comment by weiss-i-net** [comment]
Smaller example:
```python
from typing import Callable, Any

f: Callable[[int, int], Any] = lambda a, b: a @ b # error
g: Callable[[int], Any] | Callable[[int, int], Any] = lambda a, b: a @ b # no error
```

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.361, which I just published.
